### ISSUE TITLE
dyno parser mishandles unterminated block statements

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#19922]
### Steps to Reproduce

**Source Code:**

```chapel
writeln("Who are you, really?");
{
  writeln("I am just a figment of your imagination.");
```

**Compile command:**

```
$ chpl mib.chpl
$ ./mib -nl 1
$ chpl --no-dyno mib.chpl
mib.chpl:4: syntax error
```

This could be confusing, especially to a beginner.

**Associated Future Test(s):**

None yet

### Configuration Information

```
chpl version 1.27.0 pre-release (733149ca6a)

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: unset
CHPL_AUX_FILESYS: none

gcc (Ubuntu 11.2.0-7ubuntu2) 11.2.0
```

**Comment by bradcray** [comment]
I'm gonna have nightmares about this one tonight, I just know it.

**Comment by dlongnecke-cray** [comment]
Resolved by #19923. The bug was that sometimes we would create an empty error message in `yyerror` (the hook Bison uses to report errors), and the AST builder would not propagate empty error messages upwards. This meant that compilation proceeded with a bogus module instead of terminating. The fix is to have ErrorMessage differentiate between "empty" and "default constructed", where only the latter means that the error is meaningless.

**Comment by bradcray** [comment]
Thanks David!
